lib/compat.php

```diff
@@ -304,3 +304,235 @@ function gutenberg_warn_classic_about_blocks() {
+	 *
+	 * @since 4.x.x
+	 *
+	 * @param array $image_size The calculated image size width and height (in that order).
+	 * @param array $block_attributes The block attributes.
+	 * @param array $image_meta The image attachment meta data.
+	 */
+	return apply_filters( 'block_core_image_get_width_height', $image_dimensions, $block_attributes, $image_meta );
+}
+
+function gutenberg_render_block_core_image( $html, $block ) {
+		// Old post or... something's wrong.
+	if ( empty( $html ) || empty( $block['attrs'] ) ) {
+		return $html;
+	}
+
+	$defaults = array(
+		'url' => '',
+		'alt' => '',
+		'id' => 0,
+		'align' => '',
+	);
+
+	$block_attributes = wp_parse_args( $block['attrs'], $defaults );
+
+	if ( empty( $block_attributes['url'] ) ) {
+		// We don't have enough data to construct new img tag. Fall back to the existing HTML.
+		return $html;
+	}
+
+	if ( ! empty( $block_attributes['id'] ) ) {
+		$attachment_id = (int) $block_attributes['id'];
+		$image_meta = wp_get_attachment_metadata( $attachment_id );
+	} else {
+		$image_meta = null;
+		$attachment_id = 0;
+	}
+
+	$image_dimensions = gutenberg_get_image_width_height( $block_attributes, $image_meta );
+	$image_src = '';
+	$srcset = '';
+	$sizes = '';
+
+	if ( empty( $image_dimensions ) ) {
+		// We don't have enough data to construct new img tag. Fall back to the existing HTML.
+		return $html;
+	}
+
+	$image_src = $block_attributes['url'];
+
+	$image_attributes = array(
+		'src' => $image_src,
+		'alt' => empty( $block_attributes['alt'] ) ? '' : $block_attributes['alt'],
+		'width' => $image_dimensions[0],
+		'height' => $image_dimensions[1],
+	);
+
+	if ( $image_meta ) {
+		// TODO: pass `$block_attributes` to the filter.
+		$srcset = wp_calculate_image_srcset( $image_dimensions, $image_src, $image_meta, $attachment_id );
+
+		if ( ! empty( $srcset ) ) {
+			// TODO: pass `$block_attributes` to the filter. This will let themes generate better `sizes` attribute.
+			$sizes = wp_calculate_image_sizes( $image_dimensions, $image_src, $image_meta, $attachment_id );
+		}
+
+		if ( $srcset && $sizes ) {
+			$image_attributes['srcset'] = $srcset;
+			$image_attributes['sizes'] = $sizes;
+		}
+	}
+
+	/**
+	 * Filters the image tag attributes when rendering the core image block.
+	 *
+	 * @since 4.x.x
+	 *
+	 * @param array $image_attributes The (recalculated) image attributes.
+	 *                                Note: expects valid HTML 5.0 attribute names.
+	 * @param array $block_attributes The image block attributes.
+	 * @param string $html The image block HTML coming from the editor. The img tag will be replaced.
+	 */
+	$image_attributes = apply_filters( 'render_block_core_image_tag_attributes', $image_attributes, $block_attributes, $html );
+
+	$attr = '';
+	foreach ( $image_attributes as $name => $value ) {
+		// Sanitize for valid HTML 5.0 attribute names, see: https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes.
+		$name = preg_replace( '/[^a-z0-9-]+/', '', strtolower( $name ) );
```

Comment:
should we just remove it if it's not valid? seems like this could introduce an XSS error?

```html
<img on&click="alert('hi');" />
```

by removing the unwanted letters we might create attributes we shouldn't. can you think of any valid reason to pass through an attribute that's invalid?

Reply:
Hmm, good point. Will change to remove when invalid chars. The only exception (well, more of a limitation) may be for `data-*` attribute names, but thinking we can limit these to `a-zA-Z0-9-` too.

Reply:
Out of curiosity, what's the impact of letting this go through unsanitized. These are generated by plugins?

If not for WordPress' lack of sanitization in other filtered functions, I'd not be able to do fun hacks like:

https://github.com/aduth/dones/blob/0450c88429deb4546f72d20f915551ec62a33de3/functions.php#L215

Reply:
There's some prior art to work from here in `@wordpress/escape-html` :

https://github.com/WordPress/gutenberg/blob/563ef41c663f163a725c2ef7e961b11ea6981f6e/packages/escape-html/src/index.js#L66

Is there not some existing PHP function we could be using?

Reply:
> what's the impact of letting this go through unsanitized

Right, this comes from plugins, but core (often) sanitizes values in these cases. 

> s there not some existing PHP function

There are couple but not exactly the same. We should probably introduce one. As @dmsnell points out it should fail the test when unexpected chars, not strip them.

Reply:
> As @dmsnell points out it should fail the test when unexpected chars, not strip them.

Attributes stripped in 8e73feb.

Reply:
> > The only exception (well, more of a limitation) may be for data-* attribute names
> 
> If we want to conform to the HTML spec we can do just that - it's not hard

Same RegExp exists in `escape-html` as well

https://github.com/WordPress/gutenberg/blob/563ef41c663f163a725c2ef7e961b11ea6981f6e/packages/escape-html/src/index.js#L12

Reply:
> > As @dmsnell points out it should fail the test when unexpected chars, not strip them.
> 
> Attributes stripped in [8e73feb](https://github.com/WordPress/gutenberg/commit/8e73feb27668b3837aae78ba6d59fab565d734c1).

Oops, argument order was incorrect. I'll account for this in a follow-up commit which includes use of the above regular expression.

Reply:
> > > As @dmsnell points out it should fail the test when unexpected chars, not strip them.
> > 
> > 
> > Attributes stripped in [8e73feb](https://github.com/WordPress/gutenberg/commit/8e73feb27668b3837aae78ba6d59fab565d734c1).
> 
> Oops, argument order was incorrect. I'll account for this in a follow-up commit which includes use of the above regular expression.

Corrected (with new pattern) in 19dad07.

Reply:
Hm, this will "let through" a lot chars that are not present in any HTML 5.0 attribute :) 
The intention here is to "Sanitize for valid HTML 5.0 attribute names".

All attributes there (except `data-*`) use only lower case `a-z`, `0-9` and `-`. If we really want to allow `data-*` to contain all allowed chars in attribute names (this is not the same as "valid attribute names"), then we would need to make an exception only for it.

Also `$is_invalid_attribute` should probably be `$is_invalid_attribute_name`, we only check the names there.

Frankly I'd prefer to "enforce" use of only `a-z0-9-` even for the variable part of `data-*`. This is already done in other places in core, don't see a compelling reason for it to be different here.

Reply:
Updated in https://github.com/WordPress/gutenberg/pull/11973/commits/e3a99c74079627f9037551efe30149b7f5665fc0 to check for `[^a-z0-9-]` with an exception for `data-*`. Also fixed the replace image regex a bit, we want at least one white space char after `<img`.